server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -680,72 +689,156 @@ public String toString() {
+                        return;
                     }
-                    final long requestFilePosition = position;
-                    cancellableThreads.executeIO(() ->
-                        recoveryTarget.writeFileChunk(md, requestFilePosition, content, lastChunk, translogOps.get(),
+                    if (requestSeqIdTracker.getMaxSeqNo() - requestSeqIdTracker.getCheckpoint() >= maxConcurrentFileChunks) {
+                        turnAcquired.set(false);
+                        return;
+                    }
+                }
+                try {
+                    final byte[] buffer = recycledBuffers.removeFirst();
+                    final FileChunk chunk = readChunk(buffer);
+                    if (chunk == null) {
+                        if (requestSeqIdTracker.getMaxSeqNo() == requestSeqIdTracker.getCheckpoint()) {
+                            listener.onResponse(null);
+                        }
+                        recycledBuffers.addFirst(buffer);
+                        turnAcquired.set(false);
+                        return;
+                    }
+                    final long requestSeqId = requestSeqIdTracker.generateSeqNo();
+                    turnAcquired.set(false);
+                    cancellableThreads.execute(() ->
+                        recoveryTarget.writeFileChunk(chunk.md, chunk.position, chunk.content, chunk.lastChunk, translogOps.get(),
                             ActionListener.wrap(
-                                r -> requestSeqIdTracker.markSeqNoAsCompleted(requestSeqId),
-                                e -> {
-                                    error.compareAndSet(null, Tuple.tuple(md, e));
+                                r -> {
+                                    recycledBuffers.addFirst(buffer);
                                     requestSeqIdTracker.markSeqNoAsCompleted(requestSeqId);
-                                }
+                                    sendFileChunks(listener);
+                                },
+                                e -> listener.onFailure(handleErrorOnSendFiles(chunk.md, e))
                             )));
-                    position += content.length();
+                } catch (Exception e) {
+                    listener.onFailure(e);
+                    return;
                 }
-            } catch (Exception e) {
-                error.compareAndSet(null, Tuple.tuple(md, e));
-                break;
             }
         }
-        // When we terminate exceptionally, we don't wait for the outstanding requests as we don't use their results anyway.
-        // This allows us to end quickly and eliminate the complexity of handling requestSeqIds in case of error.
-        if (error.get() == null) {
-            cancellableThreads.execute(() -> requestSeqIdTracker.waitForOpsToComplete(requestSeqIdTracker.getMaxSeqNo()));
+
+        synchronized FileChunk readChunk(final byte[] buffer) throws Exception {
```

Comment:
I don't understand why we parallelize the reading on top of a single file and then synchronize all of it. This doesn't make sense to me. I think we should build the model on top of the file and chuck head of time. ie. if we want to read with N threads in parallel then chunk the file up in N pieces and send them all in parallel. That means we must write them in the correct places on the other side as well but blocking on the read side here is not making much sense to me.

Reply:
Another option is to have a multiplexer if you want to make use of the parallelism between sending and reading. We need some kind of threadpool and a task queue for that. Once I am done reading a chunk I put it on a queue and read the next chunk. Another worker can then pick it up and send it. If the queue fills up we add more threads until we saturate. Or we do reading and sending in the same thread but notify others that another chunk can be read. But there is so much blocking going on here I feel like we didn't make the right design decisions?